File: Programmers/Level_2/Swift_Code/Swift_Code/행렬의곱셈.swift
//
//  행렬의곱셈.swift
//  Swift_Code
//
//  Created by InJe Choi on 2022/03/11.
//
/// indices를 쓰면 안전하게 배열의 범위를 전체 조회할 수 있다
/// 배열 범위 밖을 참조해서 터질 일이 없음
/// answer 배열은 행렬의 곱셈 성질을 이용하여 미리 배열의 크기를 계산해서 초기화 

import Foundation

func solution(_ arr1:[[Int]], _ arr2:[[Int]]) -> [[Int]] {
    var answer = [[Int]](repeating: Array(repeating: 0, count: arr2[0].count), count: arr1.count)
    
    for m in arr1.indices {
        for n in arr2[0].indices {
            for k in arr1[0].indices {
                answer[m][n] += arr1[m][k] * arr2[k][n]
            }
        }
    }
    
    return answer
}

// print(solution([[1, 4], [3, 2], [4, 1]], [[3, 3], [3, 3]]))

File: Programmers/Level_2/행렬의곱셈.py
'''
https://programmers.co.kr/learn/courses/30/lessons/12949

행렬 규칙을 까먹어서 좀 헤매게 된 문제
m * k, k * n 행렬이 있으면 
두 행렬을 곱했을 때 나오는 행렬은 m * n 형태이다.
그리고 행렬의 곱 연산을 하려면 무조건 k가 같은 수 이어야 한다.
고로, 결괏값이 들어갈 2차원 배열을 먼저 만들고
행렬의 성질을 이용한 곱 연산을 진행하여 최종 값을 만든다.
'''

def solution(arr1, arr2):
    answer = [[0] * len(arr2[0]) for _ in range(len(arr1))]

    for m in range(len(arr1)):
        for n in range(len(arr2[0])):
            for k in range(len(arr1[0])):
                answer[m][n] += arr1[m][k] * arr2[k][n]

    return answer
